# Data-preprocessing
## 0. Load-data

In [2]:
import pandas as pd

train_raw = pd.read_csv("data/train.csv")
test_raw = pd.read_csv("data/test.csv")
submission_raw = pd.read_csv("data/submission.csv")

train = train_raw.copy()
test = test_raw.copy()
submission = submission_raw.copy()

## 1. Feature-engineering
모든 feature의 타입이 숫자형이므로 숫자로의 변환이 필요하지 않다.

In [5]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1459 entries, 0 to 1458
Data columns (total 11 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      1459 non-null   int64  
 1   hour                    1459 non-null   int64  
 2   hour_bef_temperature    1457 non-null   float64
 3   hour_bef_precipitation  1457 non-null   float64
 4   hour_bef_windspeed      1450 non-null   float64
 5   hour_bef_humidity       1457 non-null   float64
 6   hour_bef_visibility     1457 non-null   float64
 7   hour_bef_ozone          1383 non-null   float64
 8   hour_bef_pm10           1369 non-null   float64
 9   hour_bef_pm2.5          1342 non-null   float64
 10  count                   1459 non-null   float64
dtypes: float64(9), int64(2)
memory usage: 125.5 KB


## 2. Remove-outlier
앞의 EDA 과정에서 극 소수의 대여 횟수가 나타났지만, 기존 양상에서 벗어나지 않았기 때문에 모든 값을 포함하여 먼저 진행한다.

In [6]:
train.describe()

id         hour  hour_bef_temperature  hour_bef_precipitation  \
count  1459.000000  1459.000000           1457.000000             1457.000000   
mean   1105.914325    11.493489             16.717433                0.031572   
std     631.338681     6.922790              5.239150                0.174917   
min       3.000000     0.000000              3.100000                0.000000   
25%     555.500000     5.500000             12.800000                0.000000   
50%    1115.000000    11.000000             16.600000                0.000000   
75%    1651.000000    17.500000             20.100000                0.000000   
max    2179.000000    23.000000             30.000000                1.000000   

       hour_bef_windspeed  hour_bef_humidity  hour_bef_visibility  \
count         1450.000000        1457.000000          1457.000000   
mean             2.479034          52.231297          1405.216884   
std              1.378265          20.370387           583.131708   
min              0.000000           7.000000            78.000000   
25%              1.400000          36.000000           879.000000   
50%              2.300000          51.000000          1577.000000   
75%              3.400000          69.000000          1994.000000   
max              8.000000          99.000000          2000.000000   

       hour_bef_ozone  hour_bef_pm10  hour_bef_pm2.5        count  
count     1383.000000    1369.000000     1342.000000  1459.000000  
mean         0.039149      57.168736       30.327124   108.563400  
std          0.019509      31.771019       14.713252    82.631733  
min          0.003000       9.000000        8.000000     1.000000  
25%          0.025500      36.000000       20.000000    37.000000  
50%          0.039000      51.000000       26.000000    96.000000  
75%          0.052000      69.000000       37.000000   150.000000  
max          0.125000     269.000000       90.000000   431.000000

가시성에 대한 값의 분산이 583으로 다른 변수들에 비해 너무 크게 나타나므로 단위를 조정해준다.

In [8]:
train['hour_bef_visibility'] = train['hour_bef_visibility']/10
train.describe()

id         hour  hour_bef_temperature  hour_bef_precipitation  \
count  1459.000000  1459.000000           1457.000000             1457.000000   
mean   1105.914325    11.493489             16.717433                0.031572   
std     631.338681     6.922790              5.239150                0.174917   
min       3.000000     0.000000              3.100000                0.000000   
25%     555.500000     5.500000             12.800000                0.000000   
50%    1115.000000    11.000000             16.600000                0.000000   
75%    1651.000000    17.500000             20.100000                0.000000   
max    2179.000000    23.000000             30.000000                1.000000   

       hour_bef_windspeed  hour_bef_humidity  hour_bef_visibility  \
count         1450.000000        1457.000000          1457.000000   
mean             2.479034          52.231297           140.521688   
std              1.378265          20.370387            58.313171   
min              0.000000           7.000000             7.800000   
25%              1.400000          36.000000            87.900000   
50%              2.300000          51.000000           157.700000   
75%              3.400000          69.000000           199.400000   
max              8.000000          99.000000           200.000000   

       hour_bef_ozone  hour_bef_pm10  hour_bef_pm2.5        count  
count     1383.000000    1369.000000     1342.000000  1459.000000  
mean         0.039149      57.168736       30.327124   108.563400  
std          0.019509      31.771019       14.713252    82.631733  
min          0.003000       9.000000        8.000000     1.000000  
25%          0.025500      36.000000       20.000000    37.000000  
50%          0.039000      51.000000       26.000000    96.000000  
75%          0.052000      69.000000       37.000000   150.000000  
max          0.125000     269.000000       90.000000   431.000000

In [13]:
train.value_counts('hour_bef_precipitation')

hour_bef_precipitation
0.0    1411
1.0      46
dtype: int64

비가 내렸는지에 대한 데이터를 확인해본 결과 과편향되었으므로 따로 처리를 해주거나 쓰지 않는 것이 좋아보인다.

## 3. Train -> Train/Validation

In [12]:
from sklearn.model_selection import train_test_split

X = train.drop(['id', 'count'], axis=1)
y = train['count']

X_train, X_valid, y_train, y_valid = train_test_split(X,y, test_size=0.1, shuffle=False)

#데이터 shape 확인
print(f"X_train.shape : {X_train.shape}")
print(f"y_train.shape : {y_train.shape}")
print(f"X_valid.shape : {X_valid.shape}")
print(f"y_valid.shape : {y_valid.shape}")

X_train.shape : (1313, 9)
y_train.shape : (1313,)
X_valid.shape : (146, 9)
y_valid.shape : (146,)
